### PR TITLE
Devise

### PR DESCRIPTION
**Pull request by hammad-Ikhlaq-7vals**
Made Company Model and User Devise. Make Migrations.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -28,6 +28,8 @@ gem 'jbuilder', '~> 2.7'
 # Reduces boot times through caching; required in config/boot.rb
 gem 'bootsnap', '>= 1.4.2', require: false
 
+gem 'devise'
```
Comment (mujeeb-arshad) [inline Gemfile]: add version of the gem

File: app/models/company.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+class Company < ApplicationRecord
+  has_many users
```
Comment (mujeeb-arshad) [inline app/models/company.rb]: users -> :users

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (mujeeb-arshad) [inline app/models/user.rb:5]: add belongs_to :company

File: app/views/companies/confirmations/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
+
+<%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "companies/shared/error_messages", resource: resource %>
```
Comment (mujeeb-arshad) [inline app/views/companies/confirmations/new.html.erb]: internationalization

File: app/views/companies/confirmations/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email", value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email) %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Resend confirmation instructions" %>
```
Comment (mujeeb-arshad) [inline app/views/companies/confirmations/new.html.erb]: internationalization. Do it everywhere

File: app/views/companies/shared/_error_messages.html.erb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<% if resource.errors.any? %>
+  <div id="error_explanation">
+    <h2>
+      <%= I18n.t("errors.messages.not_saved",
```
Comment (mujeeb-arshad) [inline app/views/companies/shared/_error_messages.html.erb]: - convert that to single line.
- no need to add I18n.

File: app/views/devise/shared/_error_messages.html.erb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<% if resource.errors.any? %>
+  <div id="error_explanation">
+    <h2>
+      <%= I18n.t("errors.messages.not_saved",
```
Comment (mujeeb-arshad) [inline app/views/devise/shared/_error_messages.html.erb]: .

File: config/database.yml [hunk 8]
```diff
@@ -14,7 +14,7 @@ default: &default
   encoding: utf8mb4
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   username: root
-  password: password
+  password: 7vals
```
Comment (mujeeb-arshad) [inline config/database.yml]: add this file to git ignore

File: db/schema.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# This file is auto-generated from the current state of the database. Instead
+# of editing this file, please use the migrations feature of Active Record to
```
Comment (mujeeb-arshad) [inline db/schema.rb:2]: add this file to git ignore as well

File: test/fixtures/companies.yml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (mujeeb-arshad) [inline test/fixtures/companies.yml]: delete this file

File: test/fixtures/users.yml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+# This model initially had no columns defined. If you add columns to the
+# model remove the '{}' from the fixture names and add the columns immediately
+# below each fixture, per the syntax in the comments below
```
Comment (mujeeb-arshad) [inline test/fixtures/users.yml]: delete this file

File: test/models/company_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class CompanyTest < ActiveSupport::TestCase
```
Comment (mujeeb-arshad) [inline test/models/company_test.rb]: delete this file

File: test/models/user_test.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
```
Comment (mujeeb-arshad) [inline test/models/user_test.rb]: delete this file

**Comment by mujeeb-arshad** [comment]
cycle 1

File: app/views/companies/passwords/edit.html.erb [hunk 14]
```diff
@@ -0,0 +1,25 @@
+<h2>Change your password</h2>
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render I18n.t"companies/shared/error_messages", resource: resource %>
+  <%= f.hidden_field :reset_password_token %>
+
+  <div class="field">
+    <%= f.label :password, "New password" %><br />
```
Comment (muhammad-basil) [inline app/views/companies/passwords/edit.html.erb]: t('users.new_password')

File: app/views/companies/registrations/edit.html.erb [hunk 15]
```diff
@@ -23,16 +23,16 @@
 
   <div class="field">
     <%= f.label :password_confirmation %><br />
-    <%= f.password_field :password_confirmation, autocomplete: "new-password" %>
+    <%= f.password_field :password_confirmation, autocomplete: t :new_password %>
```
Comment (muhammad-basil) [inline app/views/companies/registrations/edit.html.erb]: revert

File: config/locales/en.yml [hunk 16]
```diff
@@ -31,3 +31,17 @@
...
+  change_my_password: "Change my password"
+  confirm_new_password: "Confirm new password"
+  new_password: "new-password"
+  forget_password: "Forgot your password?"
+  email: "email"
+  send_reset_password_instuctions: "Send me reset password instructions"
+  update: "Update"
+  sign_up: "Sign up"
+  login: "Log in"
+  resend_confirmation_instructions: "Resend confirmation instructions"
```
Comment (muhammad-basil) [inline config/locales/en.yml]: - remove 33 34 35 37 40 
- move update, email, to shared block
- remaining to users block

File: app/views/devise/sessions/new.html.erb [hunk 17]
```diff
@@ -19,8 +19,8 @@
   <% end %>
 
   <div class="actions">
-    <%= f.submit "Log in" %>
+    <%= f.submit t :login %>
   </div>
 <% end %>
 
-<%= render "devise/shared/links" %>
+<%= render t :shared_path_of_links %>
```
Comment (muhammad-basil) [inline app/views/devise/sessions/new.html.erb]: revert

File: app/views/devise/sessions/new.html.erb [hunk 18]
```diff
@@ -3,12 +3,12 @@
 <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
   <div class="field">
     <%= f.label :email %><br />
-    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+    <%= f.email_field :email, autofocus: true, autocomplete: t :email %>
```
Comment (muhammad-basil) [inline app/views/devise/sessions/new.html.erb]: t('shared.email')

**Pull request merged**